Alex_RTOS_Sailboat/RTOS_framework_v1/src/sail_gps.c

```diff
@@ -0,0 +1,344 @@
+			portMAX_DELAY);                          /* Wait time does not expire */
+			
+		taskENTER_CRITICAL();
+		watchdog_counter |= 0x01;
+		taskEXIT_CRITICAL();
+
+		GPS_On();
+
+		running_task = eReadWeatherSensor; //Replace this with gps, see tasksinit.c/.h
+
+		NMEA_GenericMsg msg;
+
+		if (GPS_RxMsg(&msg) == STATUS_OK) {
+			loop_cnt++;
+			//DEBUG_Write("\nStatus OK\n");
+			//DEBUG_Write("loop count %d\r\n", loop_cnt);
+			//store msg into msg array at index corresponding to msg type
+			GPS_data.msg_array[msg.type] = msg;
+			//add type to msg type sum to keep track of the saved messages
+			GPS_data.msg_type_sum += msg.type;
+
+			DEBUG_Write("\nAll fields obtained... Going to sleep...\r\n");
+			GPS_data.msg_type_sum = 0;
+
+			//store weather station data into appropriate structs
+			assign_gps_readings();
+			//check if waypoint was reached and affect as necessary
+			DEBUG_Write("checking waypoint...\r\n");
+			check_waypoint_state();
+
+			//Necessary??????????????
+			DEBUG_Write("processing wind...\r\n");
+			//calculate wind parameters
+			process_wind_readings();
+			DEBUG_Write("processing heading...\r\n");
+			//calculate heading parameters
+			process_heading_readings();
+
+			//put thread to sleep until a specific tick count is reached
+			vTaskDelay(read_gps_delay);
+		}
+	}
+}
+
+void GPS_Sleep_Sec(unsigned time_sec) {
+	GPS_Disable();
+	//put thread to sleep for number of ticks specified
+	vTaskDelay(time_sec * configTICK_RATE_HZ);
+}
+
+void GPS_On(void) {
+	GPS_Init();
+	GPS_Enable();
+}
+
+enum status_code GPS_Init(void)
+{
+	// Return if the module has already been initialized
+	if (init_flag) {
+		return STATUS_ERR_ALREADY_INITIALIZED;
+	}
+
+	// Initialize NMEA channel
+	switch (NMEA_Init(NMEA_GPS)) {
+	case STATUS_OK:// Initialization complete, continue
+		break; 
+	case STATUS_ERR_ALREADY_INITIALIZED:	// Already initialized, not a problem
+		break;
+	default:
+		DEBUG_Write_Unprotected("NMEA module could not be initialized!\n");
+		return STATUS_ERR_DENIED;
+	}
+
+	// Set the initialization flag
+	init_flag = true;
+
+	return STATUS_OK;
+}
+
+
+enum status_code GPS_Enable(void)
+{
+	// Return if the module hasn't been initialized
+	if (!init_flag) {
+		return STATUS_ERR_NOT_INITIALIZED;
+	}
+
+	// Return if the receiver cannot be started
+	if (NMEA_Enable(NMEA_GPS) != STATUS_OK) {
+		//DEBUG_Write("NMEA receiver could not be started!\r\n");
+		return STATUS_ERR_DENIED;
+	}
+
+	return STATUS_OK;
+}
+
+
+enum status_code GPS_Disable(void)
+{
+	// Return if the module hasn't been initialized
+	if (!init_flag) {
+		return STATUS_ERR_NOT_INITIALIZED;
+	}
+
+	// Return if the receiver cannot be disabled
+	if (NMEA_Disable(NMEA_GPS) != STATUS_OK) {
+		return STATUS_ERR_DENIED;
+	}
+
+	return STATUS_OK;
+}
+
+
+/*identifies the prefix from the NMEA message and assigns it to the message type*/
+bool get_NMEA_type(eNMEA_TRX_t* type) {
```

Comment:
Would it make more sense to but this in the NMEA.c module?

Reply:
Yep, I totally missed that. Added it to the NMEA module.

Reply:
*Note: It was like that before because it was using the global message buffer from the gps files. So, I just changed it so that the method accepts a message buffer as another argument.